src/odoc.ml

```diff
@@ -157,19 +157,37 @@ let css_file ~doc_dir = doc_dir ++ "odoc.css"
 
 let toplevel_index ~doc_dir = doc_dir ++ "index.html"
 
-let setup_library_rules sctx (lib : Library.t) ~dir ~modules ~mld_files
+let setup_library_rules sctx (lib : Library.t) ~dir ~scope ~modules ~mld_files
       ~requires ~(dep_graphs:Ocamldep.Dep_graph.t Ml_kind.Dict.t) =
-  let doc_dir = SC.Doc.dir sctx (dir, lib) in
-  let obj_dir = Lib.lib_obj_dir dir lib in
-  let lib_unique_name = SC.unique_library_name sctx (Lib.internal (dir, lib)) in
+  let doc_dir = SC.Doc.dir sctx lib in
+  let obj_dir, lib_unique_name =
+    match Lib.DB.find (Scope.libs scope) lib.name with
+    | Error Not_found -> assert false
+    | Error (Hidden _) ->
+      (Utils.library_object_directory ~dir lib.name,
```

Comment:
Would you mind explaining this case a bit? I'm not sure when an internal lib would be hidden. Also, it's still unsatisfying to go outside of the lib interface to get the object dir in that case.

Reply:
It's for libraries that are optional with unavailable dependencies.

> Also, it's still unsatisfying to go outside of the lib interface to get the object dir in that case.

Indeed. I suppose we could keep the `Info.t` in this case

Reply:
done